components/canvas/canvas_data.rs

```diff
@@ -470,12 +471,44 @@ impl<'a> CanvasData<'a> {
             return; // Paint nothing if gradient size is zero.
         }
 
-        let draw_rect = Rect::new(
-            rect.origin,
-            self.backend
-                .size_from_pattern(&rect, &self.state.fill_style)
-                .unwrap_or(rect.size),
-        );
+        let draw_rect = match &self.state.fill_style {
+            Pattern::Raqote(crate::raqote_backend::Pattern::Surface(pattern)) => {
+                let pattern_rect = Rect::new(Point2D::origin(), pattern.size());
+                let mut draw_rect = rect.intersection(&pattern_rect).unwrap_or(Rect::zero());
+
+                let limitx = |draw_rect: &mut Rect<f32>, pattern_rect: &Rect<f32>| {
+                    if draw_rect.size.width > pattern_rect.size.width {
+                        draw_rect.size.width = pattern_rect.size.width;
+                    }
+                };
+                let limity = |draw_rect: &mut Rect<f32>, pattern_rect: &Rect<f32>| {
+                    if draw_rect.size.height > pattern_rect.size.height {
+                        draw_rect.size.height = pattern_rect.size.height;
+                    }
+                };
```

Comment:
I think we can replace the use of these with `draw_rect.size.width = draw_rect.size.width.min(pattern_rect.size.width)` and `draw_rect.size.height = draw_rect.size.height.min(pattern_rect.size.height)`

Reply:
Yeah, that's better! Thanks